### PR TITLE
feat : add binary_exponentiation file bit_manipulation

### PR DESCRIPTION
**Pull request by sishir2001**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: Feel free to comment and modify

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1848"></a>

### CODE REVIEW

File: bit_manipulation/binary_exponentiation.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
+/**
+ * @bit_manipulation
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:2]:
```suggestion
 * @file
```

File: bit_manipulation/binary_exponentiation.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ * @brief Binary exponentiation for calculating power
+ * Implemented this algorithm :https://cp-algorithms.com/algebra/binary-exp.html
+ * Time complexity : O(log(n))
+ * Space complexity : O(1)
+ * @author Jaladi Saai Sishir(https://github.com/sishir2001)
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:7]:
```suggestion
 * @author [Jaladi Saai Sishir](https://github.com/sishir2001)
```

File: bit_manipulation/binary_exponentiation.cpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ **/
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:11]:
```suggestion
#include <iostream>  /// for IO operations
```

File: bit_manipulation/binary_exponentiation.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+            // last bit is 1
+            res *= x;
+        }
+        x *= x;  // exponentiation
+        n /= 2;  // right shifting
+    }
+    return res;
+}
+}  // namespace binary_exponentiation
+}  // namespace bit_manipulation
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:44]:
```suggestion
}  // namespace bit_manipulation

```

File: bit_manipulation/binary_exponentiation.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+void test() {
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:45]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
static void test() {
```

File: bit_manipulation/binary_exponentiation.cpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main(int argc, char *argv[]) {
+    test();
+    return 0;
+}
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:67]:
```suggestion
}

```

File: bit_manipulation/binary_exponentiation.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    assert(bit_manipulation::binary_exponentiation::binaryExponentiation(
+               6, 2) == 36);  // x = 6 n = 2 return 36
+    std::cout << "All tests passed"
+              << "\n";
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:63]:
```suggestion
/**
 * @brief Main function
 * @param argc commandline argument count (ignored)
 * @param argv commandline array of arguments (ignored)
 * @returns 0 on exit
 */
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 😄👍

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please fix [`clang-tidy`](https://github.com/TheAlgorithms/C-Plus-Plus/runs/4328417323?check_suite_focus=true) warnings. Let us know if you need any help! 🙂

File: bit_manipulation/binary_exponentiation.cpp [hunk 8]
```diff
@@ -0,0 +1,66 @@
+/**
+ * @bit_manipulation
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:2]:
```suggestion
 * @file
```

File: bit_manipulation/binary_exponentiation.cpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+ * @brief Binary exponentiation for calculating power
+ * Implemented this algorithm :https://cp-algorithms.com/algebra/binary-exp.html
+ * Time complexity : O(log(n))
+ * Space complexity : O(1)
+ * @author Jaladi Saai Sishir(https://github.com/sishir2001)
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:7]:
```suggestion
 * @author [Jaladi Saai Sishir](https://github.com/sishir2001)
```

File: bit_manipulation/binary_exponentiation.cpp [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+ * implementation
+ */
+namespace binary_exponentiation {
+/**
+ * @brief The main function implements power through binary exponentiation
+ * @param x is the base of the power
+ * @param n is the exponent for the power
+ * @returns power(x,n) i.e x^n of type long (64 bits)
+ */
+long binaryExponentiation(int x, int n) {
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:31]:
```suggestion
int64_t binaryExponentiation(int x, int n) {
```

File: bit_manipulation/binary_exponentiation.cpp [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    long res = 1;  // variable for storing the final result
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:32]:
```suggestion
    int64_t res = 1;  // variable for storing the final result
```

**Review by Panquesito7 — COMMENTED** [review]
These should resolve the warnings 🙂 just wait about 30 seconds or a minute before applying two suggestions. Apply one, wait 30 seconds, apply another one, etc..

File: bit_manipulation/binary_exponentiation.cpp [hunk 12]
```diff
@@ -0,0 +1,72 @@
+/**
+ * @file
+ * @brief Binary exponentiation for calculating power
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:3]: Provide a Wikipedia link in Markdown format (if available/possible).
If there's no Wikipedia link, add another web source for algorithm explanation. 🙂

File: bit_manipulation/binary_exponentiation.cpp [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+              << "\n";
+}
+/**
+ * @brief Main function
+ * @param argc commandline argument count (ignored)
+ * @param argv commandline array of arguments (ignored)
+ * @returns 0 on exit
+ */
+int main(int argc, char *argv[]) {
+    test();
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:70]:
```suggestion
    test();  // run self-test implementations
```

File: bit_manipulation/binary_exponentiation.cpp [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+            // last bit is 1
+            res *= x;
+        }
+        x *= x;  // exponentiation
+        n /= 2;  // right shifting
+    }
+    return res;
+}
+}  // namespace binary_exponentiation
+}  // namespace bit_manipulation
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:44]:
```suggestion
}  // namespace bit_manipulation

```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: bit_manipulation/binary_exponentiation.cpp (hunk 12 above)
Comment (sishir2001) [reply bit_manipulation/binary_exponentiation.cpp:3]: Sure !

File: bit_manipulation/binary_exponentiation.cpp [hunk 15]
```diff
@@ -0,0 +1,74 @@
+/**
+ * @file
+ * @brief Binary exponentiation for calculating power
+ * @details
+ * reference link :https://cp-algorithms.com/algebra/binary-exp.html
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:5]:
```suggestion
```

File: bit_manipulation/binary_exponentiation.cpp [hunk 16]
```diff
@@ -0,0 +1,74 @@
+/**
+ * @file
+ * @brief Binary exponentiation for calculating power
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:3]:
```suggestion
 * @brief Implementation of the [Binary exponentiation](https://cp-algorithms.com/algebra/binary-exp.html) algorithm for calculating power
```

File: bit_manipulation/binary_exponentiation.cpp [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+ * @details
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:4]: Provide a detailed description of the algorithm: what does it do, how does it work, etc..
Let us know if you need any help here or in our [**Discord community**](https://discord.gg/c7MnfGFGa6)! 😄

File: bit_manipulation/binary_exponentiation.cpp [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+#include <iostream>  /// for IO operations
+
+/**
+ * @namespace bit_manipulation
+ * @brief Bit manipulation algorithms
+ */
+namespace bit_manipulation {
+/**
+ * @namespace binary_exponentiation
+ * @brief Functions for the [binary
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:21]:
```suggestion
 * @brief Functions for the [Binary
```

File: bit_manipulation/binary_exponentiation.cpp [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+ * exponentiation](https://cp-algorithms.com/algebra/binary-exp.html)
```
Comment (Panquesito7) [inline bit_manipulation/binary_exponentiation.cpp:22]:
```suggestion
 * exponentiation](https://cp-algorithms.com/algebra/binary-exp.html) algorithm implementation
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by appgurueu** [comment]
`bit_manipulation` is the wrong place for this as no bit manipulation is being done - is there no `numeric` namespace?

**Comment by Panquesito7** [comment]
> `bit_manipulation` is the wrong place for this as no bit manipulation is being done - is there no `numeric` namespace?

Nope, there's no `numeric` namespace. We can use the [`others`](https://github.com/TheAlgorithms/C-Plus-Plus/tree/master/others) folder, though. 🙂

**Pull request closed without merging**